src/clustering_quality.jl

```diff
@@ -0,0 +1,314 @@
+ - `dist_matrix::AbstractMatrix`: a ``n×n`` pairwise distance matrix; `dist_matrix[i,j]` is the distance between ``i``-th and ``j``-th points
+
+ # Keyword arguments
+ - `quality_index::Symbol`: quality index to calculate; see below for the supported options
+ - `fuzziness::Real`: clustering fuzziness > 1
+ - `metric::SemiMetric=SqEuclidean()`: `SemiMetric` object that defines the metric/distance/similarity function
+
+When calling `clustering_quality` one can give `centers`, `assignments` or `weights` arguments by hand or provide a single `clustering` argument from which the necessary data will be read automatically.
+
+For clustering without known cluster centers the datapoints are not required, only `dist_matrix` is necessary. If given, `data` and `metric` will be used to calculate distance matrix instead.
+
+# Supported quality indices
+
+Symbols ↑/↓ are quality direction.
+- `:calinski_harabasz`: hard or fuzzy Calinski-Harabsz index (↑) returns the corrected ratio of between cluster centers inertia and within-clusters inertia
+- `:xie_beni`: hard or fuzzy Xie-Beni index (↓) returns ratio betwen inertia within clusters and minimal distance between cluster centers
+- `:davies_bouldin`: Davies-Bouldin index (↓) returns average similarity between each cluster and its most similar one, averaged over all the clusters
+- `:dunn`: Dunn index (↑) returns ratio between minimal distance between clusters and maximal cluster diameter; it does not make use of `centers` argument
+- `:silhouettes`: average silhouette index (↑), for all silhouettes use [`silhouettes`](@ref) method instead; it does not make use of `centers` argument
+Please refer to the [documentation](@ref clustering_quality) for the definitions and usage descriptions of the supported quality indices. 
+
+"""
+function clustering_quality(
+        data::AbstractMatrix{<:Real},           # d×n matrix
+        centers::AbstractMatrix{<:Real},        # d×k matrix
+        assignments::AbstractVector{<:Integer}; # n vector
+        quality_index::Symbol,
+        metric::SemiMetric=SqEuclidean()
+    )
+    d, n = size(data)
+    dc, k = size(centers)
+
+    d == dc || throw(DimensionMismatch("Inconsistent array dimensions for `data` and `centers`."))
+    (1 <= k <= n) || throw(ArgumentError("Number of clusters k must be from 1:n (n=$n), k=$k given."))
+    k >= 2 || throw(ArgumentError("Quality index not defined for the degenerated clustering with a single cluster."))
+    n == k && throw(ArgumentError("Quality index not defined for the degenerated clustering where each data point is its own cluster."))
+    for i in eachindex(assignments)
+        (assignments[i] in axes(centers, 2)) || throw(ArgumentError("Bad assignments[$i]=$(assignments[i]) is not a valid index for `data`."))
+    end
+    for i in 1:k
+        i ∉ assignments && @warn "Cluster number $(i) is empty. Clustering quality calculation may not be reliable."
+    end
```

Comment:
This is *O(N*K)* check, i.e. very expensive. We can improve it to *O(N+K)* if we scan assignments once and fill a boolean mask of observed cluster indices. Filling the mask could be done in the loop over assignments that you have above.
But ATM I'm not sure we need an explicit check.
Referring to the previous discussion -- we should have CI test cases with the gaps in the cluster indices (I think you have just added those), but if the returned results make sense (correct value is returned or reasonable error is thrown), we don't need to check in advance.

Reply:
I've checked and alas, every alorithm reacts differently to empty clusters. We have the whole range: normal val, NaN, Inf. That's why I thought warning could be for the best.